product_contract/models/sale_order.py

```diff
@@ -0,0 +1,142 @@
+            "payment_term_id": self.payment_term_id.id,
+            "fiscal_position_id": self.fiscal_position_id.id,
+            "invoice_partner_id": self.partner_invoice_id.id,
+            "line_recurrence": self.partner_invoice_id.id,
+        }
+
+    def action_create_contract(self):
+        contract_model = self.env["contract.contract"]
+        contracts = self.env["contract.contract"]
+        for rec in self.filtered("is_contract"):
+            line_to_create_contract = rec.order_line.filtered(
+                lambda r: not r.contract_id and r.product_id.is_contract
+            )
+            line_to_update_contract = rec.order_line.filtered(
+                lambda r: r.contract_id
+                and r.product_id.is_contract
+                and r
+                not in r.contract_id.contract_line_ids.mapped("sale_order_line_id")
+            )
+            contract_templates = self.env["contract.template"]
+            for order_line in line_to_create_contract:
+                contract_template = order_line.product_id.with_company(
+                    rec.company_id
+                ).property_contract_template_id
+                if not contract_template:
+                    raise ValidationError(
+                        _(
+                            "You must specify a contract "
+                            "template for '{}' product in '{}' company."
+                        ).format(order_line.product_id.name, rec.company_id.name)
+                    )
+                contract_templates |= contract_template
+            for contract_template in contract_templates:
+                order_lines = line_to_create_contract.filtered(
+                    lambda r, template=contract_template: r.product_id.with_company(
+                        r.order_id.company_id
+                    ).property_contract_template_id
+                    == template
+                )
+                contract = contract_model.create(
+                    rec._prepare_contract_value(contract_template)
+                )
+                contracts |= contract
+                contract._onchange_contract_template_id()
+                contract._onchange_contract_type()
+                order_lines.create_contract_line(contract)
+                order_lines.write({"contract_id": contract.id})
+            for line in line_to_update_contract:
+                line.create_contract_line(line.contract_id)
+        return contracts
+
+    def action_confirm(self):
+        """If we have a contract in the order, set it up"""
+        self.filtered(
+            lambda order: (order.company_id.create_contract_at_sale_order_confirmation)
+        ).action_create_contract()
+        return super(SaleOrder, self).action_confirm()
+
+    @api.depends("order_line")
+    def _compute_contract_count(self):
+        for rec in self:
+            rec.contract_count = len(
+                rec.order_line.mapped("contract_id").filtered(lambda r: r.active)
```

Comment:
```suggestion
                rec.order_line.mapped("contract_id")
```

active is built-in in all queries by the ORM